pyteal/ast/subroutine.py

```diff
@@ -116,19 +116,20 @@ def _validate_parameter_type(
                 return abi.type_spec_from_annotation(ptype)
             else:
                 raise TealInputError(
-                    "Function has parameter {} of disallowed type {}. Only the types {} are allowed".format(
-                        parameter_name, ptype, (Expr, ScratchVar, "ABI")
-                    )
+                    f"Function has parameter {parameter_name} of disallowed type {ptype}. "
+                    f"Only the types {(Expr, ScratchVar, 'ABI')} are allowed"
                 )
 
     @staticmethod
     def _arg_types_and_by_refs(
         sig: Signature,
-        annotations: Dict[str, type],
-    ) -> Tuple[
-        List[Union[Type[Expr], Type[ScratchVar], abi.TypeSpec]],
-        Set[str],
-        Dict[str, abi.TypeSpec],
+        annotations: dict[str, type],
```

Comment:
Consider converting this and others to `@classmethod`. This will allow calling via either `self.method()` in an instance method or `cls.method()` in another class method, instead of referring by `ClassName.method()`

Reply:
please note that #310 proposes to consolidate the validation logic into `_validate()` and so this function would not exist in that proposal.